config/m4/compiler.m4

```diff
@@ -184,10 +184,11 @@ AC_DEFUN([CHECK_COMPILER_FLAG],
          CFLAGS="$BASE_CFLAGS $CFLAGS $2"
          AC_COMPILE_IFELSE([$3],
                            [AC_MSG_RESULT([yes])
-                            $4],
+                            $4
+                            CFLAGS="$SAVE_CFLAGS"],
```

Comment:
shouldn't this be before `$4` ? if not how does it fix recursion?

Reply:
this is "true" branch, in case if $4 contains CHECK_COMPILER_FLAG macro - it will be checked with enabled flag, it is ok because compiler accept such flag. In "false" branch checking flag is removed before action execute because compiler is failed to compile due to incorrect flag

Reply:
seems like this really depends on how this macro is used: that false case is nested, but true case is not nested
can we avoid nesting altogether?

Reply:
we can avoid, but this issue https://github.com/openucx/ucx/issues/4129 may be easily fixed by nested ADD_COMPILER_FLAG_IF_SUPPORTED (derivative from CHECK_COMPILER_FLAG) macro.

this innovation allows to check any combination of flags supported by compiler

Reply:
per offline discussion:
- action-if-true will be evaluated with new (updated) CFLAGS
- action-if-false will be evaluated with restored CFLAGS